libs/common/ring_buffer.hpp

```diff
@@ -0,0 +1,152 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_COMMON_RING_BUFFER_HPP
+#define IROHA_COMMON_RING_BUFFER_HPP
+
+namespace iroha {
+
+    template<typename T, size_t Count> class RingBuffer final {
+    public:
+        using type = T;
+        using handle = size_t;
+
+    private:
+        static_assert(Count > 0, "Unexpected count value. It must be above 0.");
+
+        enum { kSizeType = sizeof(type) };
+        enum { kActualLimit = Count };
+        enum { kCapLimit = kActualLimit + 1 };
+        enum { kVirtualLimit = 2 * kActualLimit };
+
+        struct Node {
+            uint8_t data[kSizeType];
+        } data_[Count];
+
+        handle end_;
+        handle begin_;
+
+        inline bool handleInBound_(handle register h) const {
+            /**
+             * this code is only for debug purpose
+            **/
+            auto register vbegin = begin_ + kVirtualLimit;
+            h += kVirtualLimit;
+            return ((h >= end_) && (h < vbegin));
+        }
+
+        inline size_t incrementAndNormilize(size_t val) const {
+            return (++val % kVirtualLimit);
+        }
+
+        inline size_t handle2Position(handle h) const {
+            return (h % kActualLimit);
+        }
+
+        inline size_t size_() const {
+            assert(((begin_ - end_) % kVirtualLimit) < kCapLimit);
+            return ((begin_ - end_) % kVirtualLimit);
+        }
+
+        inline bool empty_() const {
+            return (begin_ == end_);
+        }
+
+        inline type& getItem_(Node& node) {
+            return *(type*)node.data;
+        }
+
+        inline type const& getItem_(Node const& node) const {
+            return *(type const*)node.data;
+        }
+
+        inline void destruct_(Node& node) {
+            assert(!empty_());
+
+            auto& item = getItem_(node);
+            item.~type();
+            end_ = incrementAndNormilize(end_);
+        }
+
+        template<typename...__args> inline void construct_(Node& node, __args&&... args) {
+            assert(size_() < kActualLimit);
+
+            new(node.data) type(std::forward<__args>(args)...);
+            begin_ = incrementAndNormilize(begin_);
+        }
+
+        inline Node& toNode_(handle h) {
+            assert(h < kVirtualLimit);
+            auto& node = *(Node*)&data_[handle2Position(h)];
+            return node;
+        }
+
+        inline Node const& toNode_(handle h) const{
+            assert(h < kVirtualLimit);
+            auto const& node = *(Node const*)&data_[handle2Position(h)];
+            return node;
+        }
+
+    public:
+        RingBuffer() : begin_(0ull), end_(0ull) {
+            memset(data_, 0, sizeof(data_));
+        }
+
+        ~RingBuffer() {
+            while (!empty_())
+                pop();
+        }
+
+        template<typename FuncOnAdd, typename FuncOnRemove, typename...Args>
+        void push(FuncOnAdd&& on_add, FuncOnRemove&& on_remove, Args&&...args) {
```

Comment:
the function can return the handle of newly created element:
```suggestion
        handle push(FuncOnRemove&& on_remove, Args&&...args) {
```
so `getItem_(h)` will only be computed if user needs it.

Reply:
As I told you earlier(when I told you about LTO and LTCG). Compiler and linker can inline methods and keyword 'inline' is only a  hint for compiler. In this case all of the implementation of 'push ' method will be inlined. But the inlining of the method itself will depends on size of machine code generated to form stack frame ralative to size of machine code of the function, plus it takes care of code layout to avoid ping-pong of cache line reloads.
In the 'push' implementation 'getItem_' calls twice, before on_remove and on_add execution. All the code between these calls doesn't affect 'node' pointer(Yes it will be the pointer, the effective address). So they will be optimized independently. As I told above 'getItem_' will be inlined. It this case 'getItem_' in a both places will be generated ONLY if corresponding lambda function will use it. Other way all this code except the insertion of a new elemnt and destruction the old will be skipped.
In your case with the same logic 'getItem_' generation will be skipped only if 'push' will be inlined and there will be no not optimized code, depending on the result 'handle'. Other way 'push' method will be generated as a separate function and 'getItem_' will be generated always and handle will be put into RAX.
So no, makes a 'handle ' as a result param will be not the otimization, but it can reduce the code performance.

Reply:
thanks for the explanation. also keeping `on_add` an argument will keep the interface more consistent and reduce the probability that someone stores the handle when it could be avoided.